### ISSUE TITLE
Automatically split to multiple html files according to table of contents link in original PDF?

### ISSUE THREAD

**Issue by zjplab** [pdf2htmlEX/pdf2htmlEX#52]
I want to convert a book pdf to multiple htmls with a table of contents for reference purpose. Is this achievable with pdf2htmlEX?

**Comment by stephengaito** [comment]
@zjplab I am sorry to have taken so long to get back to you (alas `pdf2htmlEX` is not my "day" job).

I am not sure what you mean by multiple html files.

I regularly process my own pdf books into multiple files with an external table of contents.
However this is _split up on a per page basis_ and for some reason (which I will eventually have to work out) the *external* table of contents is off by one page.

To have multiple html files one per page and embed all css, images, javascript, fonts and outline (table of contents) use:

```
  pdf2htmlEX --split-pages 1 <<the rest of your arguments>>
```
or more explicitly:
```
  pdf2htmlEX --embed cfijo --split-pages 1 <<the rest of your arguments>>
```
See:
```
  pdf2htmlEX --help
```
```
  --embed <string>              specify which elements should be embedded into output
  --embed-css <int>             embed CSS files into output (default: 1)
  --embed-font <int>            embed font files into output (default: 1)
  --embed-image <int>           embed image files into output (default: 1)
  --embed-javascript <int>      embed JavaScript files into output (default: 1)
  --embed-outline <int>         embed outlines into output (default: 1)
  --split-pages <int>           split pages into separate files (default: 0)
```

I hope this helps.

**Comment by HeimMatthias** [comment]
Just adding, that this might be difficult to do properly. The sheer structure of PDF bookmarks complicates this.
Bookmarks are not simply pointing to a page in the pdf, but they also add scroll-to information and scaling. This can result in a weird bookmark that tells the reader to go to page 12, and then scroll 100 points into the previous page to get to the destination. (Then of course, there are JavaScript bookmarks, which we won't get into at all).
But assuming that your bookmarks are structurally sane, it is relatively easy to parse the outline-file of pdf2htmlex with a script an use this structure to create your own html. I use this to create epubs.

**Comment by HeimMatthias** [comment]
And just to illustrate the comment above, here is a sample PDF with two titles and three different types of bookmarks:
[BookmarkExample.pdf](https://github.com/pdf2htmlEX/pdf2htmlEX/files/5628300/BookmarkExample.pdf)
Both titles are located on the first page, title 1 at the top, title 2 at the bottom.
If you convert the file with pdf2htmlEx [zipped file](https://github.com/pdf2htmlEX/pdf2htmlEX/files/5628359/BookmarkExample.zip), in the files
- the first set of bookmarks is created from the pdf-structure, these are fine.
- the third set of bookmarks uses JavaScript, these do not work with pdf2htmlEX, you actually get a warning during conversion: `Warning: unknown annotation type: 8`
- the second set of bookmarks is created by hand in Acrobat. Both are fine, but note that in the outline the second bookmark is anchored on page 2: `<a class="l" href="#pf2" data-dest-detail='[2,"XYZ",-182,1079,1]'>Title 2, at bottom of page</a></li>` -- the scroll-information `1079` pts informs the reader how far to scroll up from the bottom of page 2, and since this exceeds the height of page 2, it will scroll further up into page 1, where the title is actually located.

It just goes to show that the pdf-format is bonkers, you always need to be aware of the insane structural pitfalls it offers when you do anything with it. If you use the outline to divide the output into different "chapters" (which - to reiterate - is relatively easy to do) be aware that this might not be straightforward, especially if you created the pdf bookmarks by hand.

**Comment by HeimMatthias** [comment]
And just to finalize this :-)
Anybody trying to parse the outline and the `data-dest-detail`-attribute of the links should look at the [PDF specification](https://www.adobe.com/content/dam/acom/en/devnet/pdf/PDF32000_2008.pdf#page=374), page 366 "12.3.2.2 Explicit Destinations", where the different patterns are explained.
In the pdf2htmlEX-output, there is simply a fallback-link in the href-attribut of the anchor, that points to the page that the bookmark points to (though read above for pitfalls). The normal link behaviour is overridden with JavaScript, which interpretes the destination dictionary in the `data-dest-detail`-attribute to navigate precisely to the desired point in the output.

**Comment by ooops4** [comment]
> And just to finalize this :-)
> Anybody trying to parse the outline and the `data-dest-detail`-attribute of the links should look at the [PDF specification](https://www.adobe.com/content/dam/acom/en/devnet/pdf/PDF32000_2008.pdf#page=374), page 366 "12.3.2.2 Explicit Destinations", where the different patterns are explained.
> In the pdf2htmlEX-output, there is simply a fallback-link in the href-attribut of the anchor, that points to the page that the bookmark points to (though read above for pitfalls). The normal link behaviour is overridden with JavaScript, which interpretes the destination dictionary in the `data-dest-detail`-attribute to navigate precisely to the desired point in the output.

how can we stop from being generated this data-dest-detail attribute ? is there any flag option available ?

**Comment by HeimMatthias** [comment]
@ooops4 
(Disclaimer: I do not develop or maintain this project, I'm merely an intensive user).
I don't think there is a command line option to remove the dest-detail attribute in the outline element. But it seems a trivial task to remove this attribute in post-processing of the output.
If you do not include pdf2htmlEX.min.js in your output, this is also a functionality that will fall away, even though the attribute remains in the html-code.
Perhaps it would be clearer if you could describe what functionality you are after.
If you merely want to remove the precise bookmark location and replace it with an anchor to the top of that page, then you could pre-process the pdf-file, e.g. by exporting and reimporting the bookmarks with [pdf-tk](https://www.pdflabs.com/tools/pdftk-server/)'s dest-dump command ([export](https://www.pdflabs.com/docs/pdftk-man-page/#dest-op-dump-data-utf8), [import](https://www.pdflabs.com/docs/pdftk-man-page/#dest-op-update-info-utf8)), which strips the precise location from the bookmarks and turns them into bookmarks that point to the top of the page (though it may have other side-effects in your file.)

**Comment by ooops4** [comment]
I converted one pdf to html file. It had table of contents. So in order to not generate the sidebar, i used the flag  --process-outline 0 and it didnt generated the sidebar which is on left side. although hyperlink of table of contents is still working, but it generates this below code in html file.
```
<a class="l" href="#pf4" data-dest-detail="[4,&quot;XYZ&quot;,null,771.024,null]">
      <div class="d m1" style="border-style:none;position:absolute;left:93.543315px;bottom:1138.535430px;width:122.280000px;height:10.000000px;background-color:rgba(255,255,255,0.000001);"></div></a>
```
![image](https://user-images.githubusercontent.com/18586441/107379833-53d5c000-6b13-11eb-806e-96d472a50f29.png)

But what I want is to not generate this line -> { data-dest-detail=blahblahblah }

**Comment by HeimMatthias** [comment]
@ooops4 
This is not a link from the outline, this is a crossreference link on the pages of the pdf. There might be a clickable link on a fragment of text "see above", or "see page 4", or something along these lines, which links to a passage on page 4.
In pdf files links are constructed differently from what you expect from html code. Links are always defined as a rectangle on the page (visible or invisible) that can be clicked. The target of the link is defined as a page, the coordinates of the target on that page and optionally a scale factor. In that sense the target of a link is defined in the same way as the target of an outline entry.
This command gets translated into its equivalent with pdf2htmlEX. The links from your screenshot above has a fallback target to the id of page 4 `href="#pf4"` and a data-attribute with the full details of the link `data-dest-detail="[4,"XYZ",null,771.024,null]"` that is identical to the code in the pdf. It can be translated as `goto page 4, navigate to the following destination in rotated user space: left 0 / bottom: 771.024 points / do not adjust zoom value`. In the html output of pdf2htmlEX, this code can only be executed through JavaScript. (It is relatively easy to place anchors on the actual pages from these coordinates, I do not know why the original developer of pdf2htmlEX chose to implement it in this way).

Bottom line:

- there is no command line option to turn of processing of links. If you do not want links, remove them from the pdf-file beforehand. The easiest way to do this is to  run the file through a distiller, though this will reprocess all its contents.
- if you want to make links unclickable, just add `.l .d {display:none;}` to your css.
- again, if you really just want to remove the data-dest-detail-attribute, just clean the html in post-processing (e.g. use the following regex and replace it with nothing: `data-dest-detail='[^']+'`). It is still not clear to me what you want to achieve with this.
